pageserver/src/tenant/checks.rs

```diff
@@ -25,31 +28,46 @@ use super::storage_layer::LayerName;
 /// |       |    |   4   |    |       |
 ///
 /// If layer 2 and 4 contain the same single key, this is also a valid layer map.
+///
+/// However, if a partial compaction is still going on, it is possible that we get a layer map not satisfying the above condition.
+/// Therefore, we fallback to simply check if any of the two delta layers overlap. (See "a slow path...")
 pub fn check_valid_layermap(metadata: &[LayerName]) -> Option<String> {
     let mut lsn_split_point = BTreeSet::new(); // TODO: use a better data structure (range tree / range set?)
     let mut all_delta_layers = Vec::new();
     for name in metadata {
         if let LayerName::Delta(layer) = name {
-            if layer.key_range.start.next() != layer.key_range.end {
-                all_delta_layers.push(layer.clone());
-            }
+            all_delta_layers.push(layer.clone());
         }
     }
     for layer in &all_delta_layers {
-        let lsn_range = &layer.lsn_range;
-        lsn_split_point.insert(lsn_range.start);
-        lsn_split_point.insert(lsn_range.end);
+        if layer.key_range.start.next() != layer.key_range.end {
+            let lsn_range = &layer.lsn_range;
+            lsn_split_point.insert(lsn_range.start);
+            lsn_split_point.insert(lsn_range.end);
+        }
     }
-    for layer in &all_delta_layers {
-        let lsn_range = layer.lsn_range.clone();
-        let intersects = lsn_split_point.range(lsn_range).collect_vec();
-        if intersects.len() > 1 {
-            let err = format!(
-                "layer violates the layer map LSN split assumption: layer {} intersects with LSN [{}]",
-                layer,
-                intersects.into_iter().map(|lsn| lsn.to_string()).join(", ")
-            );
-            return Some(err);
+    for (idx, layer) in all_delta_layers.iter().enumerate() {
+        if layer.key_range.start.next() != layer.key_range.end {
+            let lsn_range = layer.lsn_range.clone();
+            let intersects = lsn_split_point.range(lsn_range).collect_vec();
+            if intersects.len() > 1 {
+                // A slow path to check if the layer intersects with any other delta layer.
+                for (other_idx, other_layer) in all_delta_layers.iter().enumerate() {
```

Comment:
one can make this non-`O(n^2)` by converting `lsn_split_point` to a `BTreeMap<Lsn, usize>`, where usize is the index in `all_delta_layers`. Then one can iter over `intersects` instead.

Reply:
Yes, probably, I can optimize it later.